### PR TITLE
Restart the container when one of the opflex device is connected duri…

### PR DESCRIPTION
**Pull request by tanyatukade**
…ng live migration

Signed-off-by: Tanya Tukade tanyatukade.123@gmail.com

### CODE REVIEW

File: pkg/controller/services.go [hunk 1]
```diff
@@ -924,6 +925,11 @@ func (cont *AciController) opflexDeviceChanged(obj apicapi.ApicObject) {
 	domName := obj.GetAttrStr("domName")
 	ctrlrName := obj.GetAttrStr("ctrlrName")
 
+	name, err := os.Hostname()
```
Comment (apoorva11029) [inline pkg/controller/services.go:928]: We can call this "controllerPodHostName" maybe?

Comment (tanyatukade) [reply pkg/controller/services.go:928]: sure, that's more helpful. Thanks

File: pkg/controller/services.go [hunk 2]
```diff
@@ -932,6 +938,13 @@ func (cont *AciController) opflexDeviceChanged(obj apicapi.ApicObject) {
 			for node, devices := range cont.nodeOpflexDevice {
 				if node == obj.GetAttrStr("hostName") {
 					for _, device := range devices {
+						if obj.GetAttrStr("hostName") == name {
+							if device.GetDn() != obj.GetDn() && device.GetAttr("state") == "connected" {
+								cont.log.Info("Found a device that is in connected state, this happens during a live migration")
+								cont.log.Info("Restarting container")
```
Comment (apoorva11029) [inline pkg/controller/services.go:944]: Add the name of the opflex device found in connected state here

Comment (apoorva11029) [reply pkg/controller/services.go:944]: Although it shouldn't matter too much, since we will lost the pod logs almost instantly

File: pkg/controller/services.go [hunk 3]
```diff
@@ -932,6 +938,13 @@ func (cont *AciController) opflexDeviceChanged(obj apicapi.ApicObject) {
 			for node, devices := range cont.nodeOpflexDevice {
 				if node == obj.GetAttrStr("hostName") {
 					for _, device := range devices {
+						if obj.GetAttrStr("hostName") == name {
+							if device.GetDn() != obj.GetDn() && device.GetAttr("state") == "connected" {
+								cont.log.Info("Found a device that is in connected state, this happens during a live migration")
```
Comment (snaiksat) [inline pkg/controller/services.go:943]: I would rephrase this something to the effect of:

Found two opflex devices associated with this node, one in connected state and another one transitioned to disconnected state.

**Comment by snaiksat** [comment]
Mostly looks good. nit - can you reword the commit message as:

Restart controller after live VM migration

When the aci-containers-controller is running inside a VM,
and that VM is live migrated, we restart the controller automatically,
to perform a complete sync, once the migration is detected to be
complete (when one of the two opflex devices is disconnected). 

This behavior can be turned OFF by configuration.

**Review by snaiksat — APPROVED** [review]

File: pkg/controller/config.go [hunk 4]
```diff
@@ -243,6 +245,8 @@ func InitFlags(config *ControllerConfig) {
 	flag.StringVar(&config.LBType, "loadbalancer", lbTypeAci, "Loadbalancer")
 	flag.IntVar(&config.MaxCSRTunnels, "max-csr-tunnels", 16, "Number of CSR tunnels")
 	flag.IntVar(&config.CSRTunnelIDBase, "csr-tunnel-id-base", 4001, "CSR starting tunnel ID")
+	flag.BoolVar(&config.RestartAfterLiveMigration, "restart-after-live-migration,omitempty", true,
```
Comment (apoorva11029) [inline pkg/controller/config.go:248]: Not sure if we need the flag here, @snaiksat ?

Comment (snaiksat) [reply pkg/controller/config.go:248]: per offline conversation, we are good on this?

Comment (apoorva11029) [reply pkg/controller/config.go:248]: Yes

**Comment by snaiksat** [comment]
May not be needed, closing for now, we can reopen if needed.
